network/src/services/dns_seeding/mod.rs

```diff
@@ -0,0 +1,161 @@
+            return Ok(());
+        }
+
+        let mut pubkey_bytes = [4u8; 65];
+        hex_decode(TXT_VERIFY_PUBKEY.as_bytes(), &mut pubkey_bytes[1..65])
+            .map_err(|err| format!("parse key({}) error: {:?}", TXT_VERIFY_PUBKEY, err))?;
+        let pubkey = PublicKey::from_slice(&pubkey_bytes)
+            .map_err(|err| format!("create PublicKey failed: {:?}", err))?;
+
+        let resolver = DnsConfig::load_default()
+            .map_err(|err| format!("Failed to load system configuration: {}", err))
+            .and_then(|config| {
+                DnsResolver::new(config)
+                    .map_err(|err| format!("Failed to create DNS resolver: {}", err))
+            })?;
+
+        let mut addrs = Vec::new();
+        for seed in &self.seeds {
+            debug!(target: "network", "query txt records from: {}", seed);
+            match resolver.resolve_record::<Txt>(seed) {
+                Ok(records) => {
+                    for record in records {
+                        match std::str::from_utf8(&record.data) {
+                            Ok(record) => match SeedRecord::decode_with_pubkey(&record, &pubkey) {
+                                Ok(seed_record) => {
+                                    let address = seed_record.address();
+                                    trace!(target: "network", "got dns txt address: {}", address);
+                                    addrs.push(address);
+                                }
+                                Err(err) => {
+                                    debug!(target: "network", "decode dns txt record failed: {:?}, {:?}", err, record);
+                                }
+                            },
+                            Err(err) => {
+                                debug!(target: "network", "get dns txt record error: {:?}", err);
+                            }
+                        }
+                    }
+                }
+                Err(_) => {
+                    if let Ok(addr) = seed.to_multiaddr() {
+                        debug!(target: "network", "DNS query failed, {} is a multiaddr", addr);
+                        addrs.push(addr);
+                    } else {
+                        warn!(target: "network", "Invalid domain name or multiaddr: {}", seed);
+                    }
+                }
+            }
+        }
+
+        debug!(target: "network", "DNS seeding got {} address", addrs.len());
+        self.network_state.with_peer_store_mut(|peer_store| {
+            for mut addr in addrs {
+                match addr.pop() {
+                    Some(Protocol::P2p(key)) => {
+                        if let Ok(peer_id) = PeerId::from_bytes(key.into_bytes()) {
+                            peer_store.add_discovered_addr(&peer_id, addr);
+                        }
+                    }
+                    _ => {
+                        debug!(target: "network", "Got addr without peer_id: {}", addr);
```

Comment:
This should be an `error` level log.

Reply:
PeerId is not required for dial a peer, but the connection still can be encrypted, just the dialer can not sure the listener is the listener (identify by ip:port).
So if worried about someone hack the network between dialer and listener here should be an `error` level log, if not I think debug is enough!